opentelemetry/src/metrics/meter.rs

```diff
@@ -265,15 +265,75 @@ impl Meter {
         ValueObserverBuilder::new(
             self,
             name.into(),
-            AsyncRunner::U64(Box::new(callback)),
+            Some(AsyncRunner::U64(Box::new(callback))),
             NumberKind::U64,
         )
     }
 
-    /// Creates a new `BatchObserver` that supports making batches of observations for
-    /// multiple instruments.
-    pub fn batch_observer(&self, callback: BatchObserverCallback) -> BatchObserver<'_> {
-        BatchObserver::new(self, AsyncRunner::Batch(callback))
+    /// Creates a new `BatchObserver` that supports making batches of observations
+    /// for multiple instruments or returns an error if instrument initialization
+    /// fails.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use opentelemetry::{global, metrics::BatchObserverResult, KeyValue};
+    ///
+    /// # fn init_observer() -> opentelemetry::metrics::Result<()> {
+    /// let meter = global::meter("test");
+    ///
+    /// meter.build_batch_observer(|batch| {
+    ///   let instrument = batch.u64_value_observer("test_instrument").try_init()?;
+    ///
+    ///   Ok(move |result: BatchObserverResult| {
+    ///     result.observe(&[KeyValue::new("my-key", "my-value")], &[instrument.observation(1)]);
+    ///   })
+    /// })?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn build_batch_observer<B, F>(&self, builder: B) -> Result<()>
+    where
+        B: Fn(BatchObserver<'_>) -> Result<F>,
+        F: Fn(BatchObserverResult) + Send + Sync + 'static,
+    {
+        let observer = builder(BatchObserver::new(self))?;
+        self.core
+            .new_batch_observer(AsyncRunner::Batch(Box::new(observer)))
```

Comment:
I'd probably just inline the call to `builder()` here, so you don't have to name the intermediate.

Reply:
Are you suggesting this specifically?
```rust
self.core
    .new_batch_observer(AsyncRunner::Batch(Box::new(builder(BatchObserver::new(
        self,
    ))?)))
```

Reply:
Yeah. Maybe the way rustfmt treats it's not actually an improvement in this case...